### PR TITLE
fix(@aws-amplify/pushnotification): fetch and cache the token if not in local

### LINKED ISSUE #2541: PushNotification.onRegister() not called at "@aws-amplify/pushnotification": "^1.0.20"
Hi, 

I'd want to use method `PushNotification.onNotificationOpened()` which seems to be available only at 1.0.20 or later. However, after upgrading "@aws-amplify/pushnotification" to 1.0.20, my `PushNotification.onRegister()` no longer got called. I no longer can print out the device token. 

I followed instructions at https://aws-amplify.github.io/docs/js/push-notifications and could confirm that the system was working with "@aws-amplify/pushnotification@1.0.19"

Anyone has successfully implemented `PushNotification.onNotificationOpened()` on react-native application? 

Thanks for your input,

My dependencies: 

```
"dependencies": {
    "@aws-amplify/analytics": "^1.2.10",
    "@aws-amplify/pushnotification": "^1.0.20",
    "aws-amplify": "^1.1.19",
    "react": "^16.7.0",
    "react-native": "^0.57.8"
  },
```

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
Related to #2852 
fixes #2541 

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pushnotification/src/PushNotification.ts [hunk 1]
```diff
@@ -76,39 +76,60 @@ export default class PushNotification {
...
         this.addEventListenerForAndroid(REMOTE_NOTIFICATION_OPENED, this.handleCampaignOpened);
         this.addEventListenerForAndroid(REMOTE_NOTIFICATION_RECEIVED, this.handleCampaignPush);
         RNPushNotification.initialize();
+
+        // check if the token is cached properly
+        if (!(await this._registerTokenCached())) {
+            const { appId } = this._config;
+            const cacheKey = 'push_token' + appId;
+            RNPushNotification.getToken((token) => {
+                logger.debug('Get the token from Firebase Serivce', token);
```
Comment (elorzafe) [inline packages/pushnotification/src/PushNotification.ts]:
```suggestion
                logger.debug('Get the token from Firebase Service', token);
```

File: packages/pushnotification/src/PushNotification.ts [hunk 2]
```diff
@@ -76,39 +76,60 @@ export default class PushNotification {
...
+    async initializeAndroid() {
         this.addEventListenerForAndroid(REMOTE_TOKEN_RECEIVED, this.updateEndpoint);
         this.addEventListenerForAndroid(REMOTE_NOTIFICATION_OPENED, this.handleCampaignOpened);
         this.addEventListenerForAndroid(REMOTE_NOTIFICATION_RECEIVED, this.handleCampaignPush);
         RNPushNotification.initialize();
+
+        // check if the token is cached properly
+        if (!(await this._registerTokenCached())) {
+            const { appId } = this._config;
+            const cacheKey = 'push_token' + appId;
```
Comment (elorzafe) [inline packages/pushnotification/src/PushNotification.ts]: this is not used

File: packages/pushnotification/src/PushNotification.ts [hunk 3]
```diff
@@ -76,39 +76,60 @@ export default class PushNotification {
...
+            RNPushNotification.getToken((token) => {
+                logger.debug('Get the token from Firebase Serivce', token);
+                // resend the token in case it's missing in the Pinpoint service
+                // the token will also be cached locally
+                this.updateEndpoint(token);
+            });
+        }
+    }
+
+    async _registerTokenCached(): Promise<boolean> {
```
Comment (elorzafe) [inline packages/pushnotification/src/PushNotification.ts]: I would rename it to `_getCachedToken` or something like that

File: packages/pushnotification/src/PushNotification.ts [hunk 4]
```diff
@@ -76,39 +76,60 @@ export default class PushNotification {
...
+
+        // check if the token is cached properly
+        if (!(await this._registerTokenCached())) {
+            const { appId } = this._config;
+            const cacheKey = 'push_token' + appId;
+            RNPushNotification.getToken((token) => {
+                logger.debug('Get the token from Firebase Serivce', token);
+                // resend the token in case it's missing in the Pinpoint service
+                // the token will also be cached locally
+                this.updateEndpoint(token);
```
Comment (elorzafe) [inline packages/pushnotification/src/PushNotification.ts]: What will happen if updateEndpoint fail? Is there a retry strategy? What do you think @powerful23 ?

File: packages/pushnotification/src/PushNotification.ts (hunk 2 above)
Comment (powerful23) [reply packages/pushnotification/src/PushNotification.ts]: Done.

**Pull request merged**